src/evolve.ts

```diff
@@ -0,0 +1,236 @@
+>;
+
+/**
+ * @example
+ * type T1 = GetValueByKey<{ A: '1', B: '2' }, 'A'>; // type T1 = "1"
+ * type T2 = GetValueByKey<{ A: '1', B: '2' }, 'B'>; // type T2 = "2"
+ * type T3 = GetValueByKey<{ A: '1', B: '2' }, 'C', '3'>; // type T3 = "3"
+ */
+type GetValueByKey<
+  T,
+  K extends PropertyKey,
+  Default = never,
+> = K extends keyof T ? T[K] : Default;
+
+/**
+ * Get the type of first parameter.
+ * But get `never` if the function take two or more required parameters.
+ * @example
+ * type Number = GetFirstParam<(arg1: number) => void>; // type Number = number
+ * type Number = GetFirstParam<(arg1: number, arg2?: number) => void>; // type Number = number
+ * type Number = GetFirstParam< (arg1: number, arg2?: number, arg3?: number) => void >; // type Number = number
+ * type Never = GetFirstParam<(arg1: number, arg2: number | undefined) => void>; // type Never = never
+ */
+type GetFirstParam<T extends AFunction> = T extends (
+  firstArg: infer Ret,
+  ...restArgs: infer Rest
+) => any
+  ? 0 extends Rest['length']
+    ? Ret
+    : never
+  : never;
+
+/**
+ * Creates an assumed `data` type from the type of the `evolver` argument.
+ * @example
+ * type Number = EvolveTarget<(arg1: number) => void>; // type Number = number
+ * type Nested = EvolveTarget<{
+ *   num: (arg1: number) => void;
+ *   obj: {
+ *    num: (arg1: number) => void
+ *   }
+ *   notFunc: undefined
+ * }>;
+ * // type Nested = {
+ * //     num?: number | undefined;
+ * //     obj?: {
+ * //         num?: number | undefined;
+ * //     } | undefined;
+ * // }
+ */
+type EvolveTarget<E> = E extends AFunction
+  ? GetFirstParam<E>
+  : E extends object
+    ? {
+        [K in keyof E as E[K] extends AFunction
+          ? K
+          : E[K] extends object
+            ? K
+            : never]?: EvolveTarget<E[K]>;
+      }
+    : never;
+
+/**
+ * Creates an assumed `evolver` type from the type of `data` argument.
+ * Note: `[T] extends [Something] ? ...` is conditional type avoiding distributivity.
+ * @example
+ * interface Data {
+ *   id: number;
+ *   quartile: Array<number>;
+ *   time?: { elapsed: number; remaining?: number };
+ * }
+ * type Nested = Evolver<Data>; // type Nested = {
+ * //  id?: ((data: number) => any) | undefined;
+ * //  quartile?: ((data: number[]) => any) | undefined;
+ * //  time?:
+ * //    | ((
+ * //        data:
+ * //          | {
+ * //              elapsed: number;
+ * //              remaining?: number | undefined;
+ * //            }
+ * //          | undefined
+ * //      ) => any)
+ * //    | {
+ * //        elapsed?: ((data: number) => any) | undefined;
+ * //        remaining?: ((data: number | undefined) => any) | undefined;
+ * //      }
+ * //    | undefined;
+ * //};
+ */
+type Evolver<T> = [T] extends [Array<any>]
+  ? (data: T) => any
+  : T extends object
+    ? {
+        [K in keyof T]?:
+          | ([T[K]] extends [Array<any>] ? never : (data: T[K]) => any)
+          | Evolver<T[K]>;
+      }
+    : never;
```

Comment:
1. You are over-complicating things! You don't need all these fancy tests (`[T] extends [...]`). These types show up in type-fest because that library is aimed at providing excellent types even for really "academic" edge-cases. We don't encounter issues and reports from users running into problems without them.
2. Notice that you need to make T `Required` before taking its value. Otherwise, the evolver functions would take `undefined` for optional props (which is not how the runtime behavior handles them).
3. You don't need to handle the array specifically; it will be folded into the object case, which is how the runtime would see things.

This version of Evolver works with all your tests:
```suggestion
type Evolver<T> = T extends object
  ? {
      readonly [K in keyof T]?:
        | ((data: Required<T>[K]) => unknown)
        | Evolver<T[K]>;
    }
  : never;
```

Reply:
1. When I was doing my trial and error, this solved a certain problem, but was not necessary with the current code. Thanks for bringing this to my attention.
2. Thank you very much! So this will prevent the following code from causing a type error!
```ts
  it('does not invoke function if `data` does not contain the key', function () {
    const data: { id?: number } = {};
    const expected = {};
    const transf = { id: add(1)};
    const result = evolve(data, transf);
    expect(result).toEqual(expected);
  });
```
3. [please pay attension] As I said [here in ❷](https://github.com/remeda/remeda/pull/454#issuecomment-1940654791), I changed implementation so that `evolver` cannot handle arrays of functions. So, I want to get type error in following pattern.
```ts
  it('can not handle function arrays.', function () {
    const expected = {
      quartile: [2, 1],
    };
    const result = evolve(
      {
        quartile: [1, 2],
      },
      {
        quartile: [add(1), add(-1)],
      }
    );
    expect(result).toEqual(expected);
    expectTypeOf(result).toEqualTypeOf<typeof expected>();
  });
```
In fact, when this test is run, it fails.